python/lsst/cell_coadds/_identifiers.py

```diff
@@ -53,17 +55,37 @@ class PatchIdentifiers:
     """Identifiers for the patch itself.
     """
 
-    band: Optional[str]
+    band: str | None
     """Name of the band, if any.
     """
 
     @classmethod
-    def from_data_id(cls, data_id: DataCoordinate) -> PatchIdentifiers:
+    def from_data_id(cls: Type[T], data_id: DataCoordinate, **kwargs: Any) -> T:
+        """Construct from a data ID."""
+        # The 'type: ignore' directives below are present because we require a
+        # a fully-expanded data ID, but this isn't embedded in the types.
+        return cls(
+            skymap=data_id["skymap"],  # type: ignore
+            tract=data_id["tract"],  # type: ignore
+            patch=Index2D(
+                x=data_id.records["patch"].cell_x, y=data_id.records["patch"].cell_y  # type: ignore
+            ),
+            band=data_id.get("band"),
+            **kwargs,
+        )
+
+
+@dataclass(frozen=True)
+class PatchIdentifiers(BasePatchIdentifiers):
+    """Struct of base identifiers for a coadd patch."""
+
+    @classmethod
+    def from_data_id(cls, data_id: DataCoordinate) -> PatchIdentifiers:  # type: ignore [override]
```

Comment:
VSCode editor is smart enough to identify the type correctly without specifying the return type, but mypy complains if left unspecified (without the #type: ignore comment of course)

Reply:
Yup, as expected; we've explicitly configured MyPy to complain about missing annotations on function signatures (we only want it to infer variable types when it can).